### DISCUSSION TITLE
Using ArduinoOTA library for SAMD51 Grand Central M4 board

### DISCUSSION THREAD

**Discussion by Bhindhiya** [JAndrassy/ArduinoOTA#142] [Q&A]
Hi,

I was able to successfully update the firmware on my Arduino MKR Zero board using the _OTEthernet_ sketch example available in this library. I used the W5500 Ethernet shield interfaced to my MKR Zero board.

I am attempting to do the same exercise using a Grand Central M4 board (SAMD51). I am able to successfully flash the sketch from the Arduino IDE through the COM port and I can see the Network Port listed in Tools -> Ports after the sketch is flashed successfully. 

When I flash the updated sketch using the Network Port, the console displays a message that the sketch was successfully flashed. However, the COM port is not detected anymore after this step when I connect the board and the board does not behave according to the updated code. A manual reset of the board does not have any effect either.

Has this been attempted previously? Any guidance w.r.t. any modifications that are required to be made to make this sketch work would be extremely helpful. Thank you for your support.

Best Regards,
Bhindhiya Raja

**Comment by JAndrassy** [comment]
I don't have a SAMD51 Arduino. The SAMD51 support was added by Blynk maintainer V. Shymanskyy. He tested it with Seeed Wio Terminal. 
https://github.com/jandrassy/ArduinoOTA/pull/104

**Reply by Bhindhiya** [reply]
Thank you for your response. Yes, I understand and have gone through PR #104 

I see a macro definition used in many places '_ARDUINO_ARCH_SAMD_'. This macro has been included in the code before the merge of the above-mentioned PR. Was the initial implementation meant to support all boards belonging to the SAMD family or only for SAMD21s?

**Reply by JAndrassy** [reply]
ARDUINO_ARCH_SAMD is defined for Arduino SAMD core. Adafruit and Seed Studio core for SAMD51 boards are based on the Arduino SAMD core and they didn't change the define.

**Comment by pscott-ams** [comment]


**Reply by JAndrassy** [reply]
> Could you show me your ArduinoOTA.cpp file please?

there is no ArduinoOTA.cpp

**Comment by Bhindhiya** [comment]
Hi @jandrassy,

Thank you for the explanation regarding the macro ARDUINO_ARCH_SAMD.

I am trying to understand the code written for an Arduino SAMD board by studying the library.

In _InternalStorage.cpp_ file, **lines 92-99:**
_#elif defined(ARDUINO_ARCH_SAMD)
    int rowSize = pageSize * 4;
    for (int i = 0; i < length; i += rowSize) {
      NVMCTRL->ADDR.reg = ((uint32_t)(address + i)) / 2;
      NVMCTRL->CTRLA.reg = NVMCTRL_CTRLA_CMDEX_KEY | NVMCTRL_CTRLA_CMD_ER;
      waitForReady();
    }_

I understand that _rowSize = (pageSize * 4)_ as each row consists of 4 pages. 

However, could you please help me understand why _NVMCTRL->ADDR.reg = ((uint32_t)(address + i))_ **/ 2**? 
a) I expect the NVMCTRL->ADDR register must consist of the _STORAGE_START_ADDRESS_ that must be erased. But why is ADDR set with the (address / 2)? 
b) Wouldn't a different address in the partitioned sketch region be erased when setting CTRLA with the Command key and Erase Row command in the next line?

**Comment by JAndrassy** [comment]
sorry I didn't code the SAMD support. it is inherited from the WiFi101OTA library.

**Comment by hydrofoor** [comment]
@Bhindhiya based on this post [https://forums.adafruit.com/viewtopic.php?p=928959#p928959](url) I believe you got around the issue you report in your original post. I am trying to achieve the same goal (upload sketch to Adafruit Grand Central M4 over ESP8266 with WifiEspAT) and hit the same problem.
I guess a tweak to the InternalStorage code is needed, perhaps to handle difference in flash layout/size, but this exceeds my knowledge/skill level. Any input is much appreciated.
thanks,
erik

**Reply by hydrofoor** [reply]
changing the way flash is erased before writing may be incorrect; check #201
